### ISSUE TITLE
A Self-Learning Modern Computer Science Curriculum

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2651]
A Self-Learning, Modern Computer Science Curriculum<br>
<br>
http://bit.ly/2nVnGcQ<br>
<br>
<br>
<br>
<div>
<h1>Introduction</h1>
<blockquote>
<p>"The great idea of constructive type theory is that there is no distinction between programs and proofs... Theorems are types (specifications), and proofs are programs" (Robert Harper)</p>
</blockquote>
<p>This is a collection of modern resources on various undergrad level computer science topics, for someone with an interest in theory. Use <a href="https://en.wikipedia.org/wiki/Library_Genesis">LibGen</a> if you can't buy these books. You don't have to do everything here, just the topics of interest to you. If for any reason you want a condensed version, watch the <em>Great Theoretical Ideas in Computer Science</em> lecture series to see what interests you in the field. Almost all the material in each resource is self contained so will cover necessary background which if you don't have see the <em>preliminaries</em>.</p>
<h2>How to Learn</h2>
<p>On Isaac Newton's iteration method to self-learn geometry:</p>
<blockquote>
<p>"He bought Descartes' Geometry and read it by himself .. when he was got over 2 or 3 pages he could understand no farther, than he began again and got 3 or 4 pages father till he came to another difficult place, than he began again and advanced farther and continued so doing till he made himself master of the whole without having the least light or instruction from anybody" (King's Cam.,Keynes MS 130.10,fol. 2/v/)</p>
</blockquote>
<p>Cal Newport has some anecdotes on how he was able to get the <a href="http://calnewport.com/blog/2008/11/25/case-study-how-i-got-the-highest-grade-in-my-discrete-math-class/">best</a> grade in his Discrete Mathematics class, and the rest of the site is full of advice on studying, how to schedule yourself and deliberate <a href="http://calnewport.com/blog/2010/01/06/the-grandmaster-in-the-corner-office-what-the-study-of-chess-experts-teaches-us-about-building-a-remarkable-life/">practice</a>. Further anecdotes exist on the importance of deep work and <a href="https://azeria-labs.com/the-importance-of-deep-work-the-30-hour-method-for-learning-a-new-skill/">deliberate</a> practice. There is a free <a href="https://www.coursera.org/learn/learning-how-to-learn">course</a> <em>Learning How to Learn</em> on coursera.org which suggest the same techniques plus Cornell style note taking and how to read a <a href="https://www.cs.jhu.edu/~jason/advice/how-to-read-a-paper.html">research paper</a>.</p>
<h1>Preliminaries</h1>
<p>Everything here is optional, you could try just starting with <em>CS 3110</em>.</p>
<h2>Intro to Computation</h2>
<p>The <a href="https://htdp.org/">book</a> <em>How to Design Programs (HtDP)</em> has proven success teaching an introduction to programming while remaining rigorous. One of it's authors is now running a public school curriculum project called <a href="https://www.bootstrapworld.org/index.shtml">Bootstrap</a> and gave a <a href="https://youtu.be/5c0BvOlR5gs">talk</a> why this style of teaching programming has the most success, and how it's curriculum transfers to other fields such as how the students writing games are actually writing differential equations, as the behavior of a game over time is the integral. If you really want to understand programming you should start with this book. If you liked HtDP, it's sequel is <a href="https://papl.cs.brown.edu/2018/index.html">PAPL</a> (original <a href="http://cs.brown.edu/courses/cs173/2012/book/">Racket</a> version also exists, but Pyret and Racket are essentially the same languages just the author was tired of parenthesis complaints from other professors). A very rigorous introduction exists as the classic MIT book <em>Structural Interpretation of Computer Programs</em> <em>(SICP)</em> but the more programming you do before reading SICP, the more you get out of it.</p>
<h2>Some Additional Optional Intros</h2>
<h3>Common Lisp Intro</h3>
<p>It's possible to do (or at least model) every course in this list using Common Lisp, such as building your own domain-specific <a href="https://www.pvk.ca/Blog/2014/03/15/sbcl-the-ultimate-assembly-code-breadboard/">machine code</a> while completing 15-213 <em>Computer Systems</em>, or hacking together <a href="https://common-lisp.net/project/fset/Site/FSet-Tutorial.html">purely</a> functional <a href="https://common-lisp.net/project/funds/funds.pdf">data structures</a>, or writing your own pattern matcher, or simulating a turing machine, ect. Of course you will do the courses using their recommended languages and tools, but in addition to this, building your own toy prototypes for the topics here in Lisp can help you learn this material as you discover it for yourself with active learning. Another shill for Common Lisp as a good language choice is it hasn't changed much in decades, so an old library you found that hasn't been updated in 15 years will still work. There is also the joy of programming Lisp as a <a href="http://stevelosh.com/blog/2018/08/a-road-to-common-lisp/#lisp-as-a-system">system</a>.</p>
<h3>Little Schemer Series</h3>
<p>The <a href="http://www.ccs.neu.edu/home/matthias/BTLS/">Little Schemer series</a> books are a Q&amp;A format/Socratic method for learning the basics of computation (read the Preface of each book). You can do the Little Schemer with pencil and paper in a weekend though the authors recommend at least 3 sittings. The first in the series is <em>The Little Schemer</em> which teaches you to think recursively. The second is the <em>Seasoned Schemer</em> covering higher-order functions, which will help understand calculus. The differential operator is a higher-order function: given a function on the real line it yields its first derivative as a function on the real line.</p>
<ul>
<li>
<a href="http://www.ccs.neu.edu/home/matthias/BTML/">(Book) The Little MLer - Matthias Felleisen, Daniel P. Friedman</a>
<ul>
<li>Self contained but will make much more sense if you've already done the first book, <em>The Little Schemer</em> covering cons, lists and recursion.</li>
<li>Teaches you to think recursively, and provides an introduction to the principles of types, computation, and program construction.</li>
<li>Applies to any functional language, almost all the exercises can be done in OCaml or another language with a small amount of syntactic adjustment.</li>
</ul>
</li>
</ul>
<h3>Carnegie Mellon University</h3>
<p>Students with direct entry to this CMU course already have a background in programming. Good information here on style, design and efficiency, with an introduction to <a href="https://www.cs.cmu.edu/~112/notes/notes-limits-of-computation.html">complexity</a> otherwise you would learn all of this in HtDP.</p>
<h2>Elementary Math</h2>
<blockquote>
<p>"So I went to Case, and the Dean of Case says to us, it's a all men's school, "Men, look at, look to the person on your left, and the person on your right. One of you isn't going to be here next year; one of you is going to fail." So I get to Case, and again I'm studying all the time, working really hard on my classes, and so for that I had to be kind of a machine. In high school, our math program wasn't much, and I had never heard of calculus until I got to college. But the calculus book that we had was (in college) was great, and in the back of the book there were supplementary problems that weren't assigned by the teacher. So this was a famous calculus text by a man named George Thomas (<a href="https://www.amazon.com/gp/customer-reviews/R1HJTYVOE94R83/ref=cm_cr_dp_d_rvw_ttl?ie=UTF8&amp;ASIN=0201077795">second edition</a>), and I mention it especially because it was one of the first books published by Addison-Wesley, and I loved this calculus book so much that later I chose Addison-Wesley to be the publisher of my own book. Our teacher would assign, say, the even numbered problems, or something like that (from the book). I would also do the odd numbered problems. In the back of Thomas's book he had supplementary problems, the teacher didn't assign the supplementary problems; I worked the supplementary problems. I was scared I wouldn't learn calculus, so I worked hard on it, and it turned out that of course it took me longer to solve all these problems than the kids who were only working on what was assigned, at first. But after a year, I could do all of those problems in the same time as my classmates were doing the assigned problems, and after that <em>I could just coast in mathematics, because I'd learned how to solve problems</em>" (<a href="https://github.com/kragen/knuth-interview-2006">Don Knuth</a> )</p>
</blockquote>
<p>I bought George Thomas' <em>Calculus and Analytical Geometry</em> <a href="https://www.amazon.com/Calculus-Analytic-Geometry-3rd-Mathematics/dp/B0000CKUFL">3rd edition</a> for around $12 off Thrift Books and can see why Knuth liked this book so much as Thomas carefully goes through each proof describing what is happening. The second or classic edition is the one Knuth used and was <a href="https://www.amazon.com/Calculus-Analytic-Geometry-Supplementary-Problems/dp/0201077795">reprinted</a> recently. There's a review of trigonometry in the 3rd edition as well.</p>
<h3>Nonexistent Math Background</h3>
<p>The Gelfand Correspondence <a href="http://gcpm.rutgers.edu/former_description.html">Program</a> in Mathematics existed in the 90s, where students were sent books and could mail back assignments to be graded and coached. These <a href="http://gcpm.rutgers.edu/books.html">books</a> are still around and are excellent taking a problem solving approach on every page. <em>Algebra</em> and <em>Trigonometry</em> are both recommended. While you go through any of these books try a survey of The <a href="https://betterexplained.com/ebook/math/">Better Explained</a> Math and Calculus series books, which explain basic math such as the natural logarithm. Gilbert Strang also has a lecture series <a href="https://ocw.mit.edu/resources/res-18-005-highlights-of-calculus-spring-2010/">Highlights of Calculus</a>. Cambridge also offers advice for incoming undergrads on readable <a href="http://www.maths.cam.ac.uk/documents/reading-list.pdf/">texts</a>.</p>
<h3>Weak/Forgotten Math Background</h3>
<p>Start with <a href="https://en.wikipedia.org/wiki/Po-Shen_Loh">Po-Shen Loh's</a> new site <a href="https://www.expii.com/">Expii</a>.com which presents topics in a brief intuitive manner, then feeds you problems in that domain and makes decisions based on your answers if you need more practice problems. The Advanced Topics have all kinds of great tutorials on CS Theory as well as Number Theory.</p>
<h3>Calculus With Theory</h3>
<p>Optional, Thomas' text will teach you the theory of calculus, though doing the exercises of Apostol's <em>Calculus I</em> is a great way to build <a href="https://en.wikipedia.org/wiki/Mathematical_maturity">mathematical maturity</a>. This alternative Calculus w/Theory lecture <a href="https://www.youtube.com/playlist?list=PLzdiPTrEWyz4rKFN541wFKvKPSg5Ea6XB">series</a> only uses a simple algebraic framework, eliminating limits/infinity. Linear algebraic ideas and the Discrete Calculus are both highly relevant to the material here. There is also a <a href="https://www.openlearning.com/courses/algebraic-calculus-one">full course</a> available.</p>
<h1>Functional Programming</h1>
<h2>Principles of Functional Programming</h2>
<p>The 2012 version by <a href="http://dlicata.web.wesleyan.edu/">Dan Licata</a> has the best lecture notes, optionally combine with most <a href="http://www.cs.cmu.edu/~15150/lect.html">recent</a> course notes. Robert Harper keeps a <a href="https://existentialtype.wordpress.com/2012/08/17/intro-curriculum-update/">blog</a> and a follow up <a href="https://existentialtype.wordpress.com/2011/03/21/the-dog-that-didnt-bark/">post</a> on the success of teaching this course. SML is used because it's syntax and module system is simple, much like Scheme's easy syntax was once preferred for introductory courses so the complexity of the language gets out of the way and you learn the fundamentals.</p>
<h2>Intro to Programming using OCaml</h2>
<p>This Cornell OCaml course is totally self contained with it's own <a href="http://www.cs.cornell.edu/courses/cs3110/2018fa/textbook/">textbook</a>. You will want to archive the entire course locally using <a href="https://www.guyrutenberg.com/2014/05/02/make-offline-mirror-of-a-site-using-wget/">wget</a>. The release code for the assignments wget them <a href="https://www.cs.cornell.edu/courses/cs3110/2018sp/a5/a5.zip">here</a> and change URL to /a4/a4.zip, /a3/a3.zip, they range a0.zip - a5.zip</p>
<ul>
<li>
<a href="https://www.cs.cornell.edu/courses/cs3110/2018sp/">(Full Course) CS 3110 Data Structures and Functional Programming</a>
<ul>
<li>The free textbook is essentially the lectures, the <a href="https://www.cs.cornell.edu/courses/cs3110/2018sp/">notes</a> have recommended chapters for additional books like <em>Real World Ocaml</em>
</li>
<li>Introduction to Coq, you can extract OCaml from Coq, F* or from <a href="http://www.cs.ru.nl/~gondelman/toolchain.pdf">Why3</a>
</li>
<li>No prereqs in programming required, if you get lost on something directly look it up in <a href="http://greenteapress.com/thinkocaml/">this</a> free book <em>How to Think Like a (functional) Computer Scientist</em>
<ul>
<li>See this <a href="https://www.youtube.com/watch?v=-J8YyfrSwTk">talk</a> Effective ML how to properly write interfaces, error handling</li>
<li>You can even program hardware such as a <a href="https://www.youtube.com/watch?v=X1cgRXhpQLY">FPGA</a> in OCaml using this <a href="https://github.com/janestreet/hardcaml">library</a>
</li>
<li>Or websites using <a href="https://bucklescript.github.io/">bucklescript</a>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<h1>Algebra</h1>
<p>It is possible in a functional language like ML to do algebra with types, proving two types are isomorphic with the desired properties of reflexivity, symmetry, and transitivity. It's also possible to abstract Lists and Trees into polynomials, as every polynomial looks like a sum of terms. As you will learn in 15-150 <em>Principles of Functional Programming</em> "most functional datastructures have constant time access near the outer layer of their structure, ie: the head of a list or the root of a tree. However, access at some random point inside the structure is typically linear since looking at some element of a list is linear in the length of the list, and looking at some element of a tree is linear in the depth of the tree. Datatype derivatives allow constant time access to the entire structure."</p>
<h2>Linear Algebra</h2>
<ul>
<li>
<a href="http://cs.brown.edu/courses/cs053/current/lectures.htm">(Full Course) CS053 - Coding the Matrix</a>
<ul>
<li>Covers interesting applications like what was shown in <em>Great Theoretical Ideas in Computer Science</em> when you learned how a parity bit works.</li>
<li>No formal prerequisites except assumes you know how to do basic proofs</li>
</ul>
</li>
<li>
<a href="http://www.springer.com/gp/book/9780387940991">(Book) Linear Algebra: An Introduction to Abstract Mathematics - Robert J. Valenza</a>
<ul>
<li>This free <a href="https://www.math.ucdavis.edu/~anne/linear_algebra/index.html">book</a> <em>Linear Algebra - As an Introduction to Abstract Mathematics</em> from UC Davis, is an excellent companion to Valenza's text</li>
<li>These YouTube presentations <a href="https://www.youtube.com/playlist?list=PLZHQObOWTQDPD3MizzM2xVFitgF8hE_ab">here</a> help with understanding the topics on a geometric level. Watch the <a href="https://youtu.be/kjBOesZCoqc?list=PLZHQObOWTQDPD3MizzM2xVFitgF8hE_ab">Essence of Linear Algebra</a> to see what this means
<ul>
<li>Other Resource: Sheldon Axler has <a href="https://www.youtube.com/playlist?list=PLGAnmvB9m7zOBVCZBUUmSinFV0wEir2Vw">recorded lectures</a> for his book <em>Linear Algebra Done Right</em> aka Down With Determinants!
<ul>
<li>The exercises in Axler's book are either too easy, sometimes near impossible, but the presentation of material with vectors explained first before determinants is preferred</li>
</ul>
</li>
<li>Yet Another Resource: NJ Wildberger has a Linear Algebraic Geometry series of <a href="https://www.youtube.com/watch?v=yAb12PWrhV0&amp;list=PLIljB45xT85BhzJ-oWNug1YtUjfWp1qAp">lectures</a>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>Abstract Algebra</h2>
<ul>
<li>
<a href="https://www.math.upenn.edu/~ted/371F14/hw-371SchedTab.html">(Full Course) Math-371 Abstract Algebra</a>
<ul>
<li>Has recorded lectures which are essential as you can quickly get lost in the texts at this level of abstraction</li>
<li>Uses readings from <a href="https://www.math.upenn.edu/~ted/371F14/info.html#texts">three texts</a>
<ul>
<li>There is also this <a href="http://www.extension.harvard.edu/open-learning-initiative/abstract-algebra">Harvard course</a> with excellent recorded lectures and uses readings from <em>Algebra - M. Artin</em> (blue book) as a great compliment to Math-371 lectures</li>
</ul>
</li>
<li>The <a href="http://www.math.hawaii.edu/~lee/algebra/history.html">history</a> of Abstract Algebra by Prof Lee Lady</li>
</ul>
</li>
</ul>
<h1>Discrete Mathematics</h1>
<p>Resources to learn undergrad Discrete Mathematics.</p>
<h2>Introduction to Pure Mathematics</h2>
<ul>
<li>
<a href="https://infinitedescent.xyz/">(Book) An Infinite Descent into Pure Mathematics</a>
<ul>
<li>Collection of excellent lecture notes used in CMU's 15-151/21-127 Concepts of Mathematics class</li>
<li>Curriculum based on findings in evidenced based teaching</li>
<li>Author's motivations for not including answers are explained <a href="https://infinitedescent.xyz/20180214-exercises-ibl/">here</a>: "people learn more when they discover something for themselves than they do if someone tells them about it."
<ul>
<li>Alternatively try the <a href="https://archive.org/details/B-001-002-135">book</a> <em>Concrete Mathematics</em> for a complete description of proof by induction, summation notation, generating functions, ect., with answers included for self-study.</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>Discrete Math with Standard ML</h2>
<p>The merging of proof and program is what makes this a great book. If you can't get this book <em>Discrete Mathematics</em>, by L. Lov&aacute;sz, J. Pelik&aacute;n, and K. Vesztergombi is normally <a href="http://www.math.cmu.edu/~lohp/2013-228.shtml">used</a> or just skip to 15-251 <em>Great Theoretical Ideas in CS</em>. If you can finish this book you've essentially done most of 15-150 <em>Principles of Functional Programming</em> (such as proving programs correct) and most of 15-215 <em>Great Theoretical Ideas in Computer Science</em>.</p>
<ul>
<li>
<a href="https://cs.wheaton.edu/~tvandrun/dmfp/">(Book/Lectures) Discrete Mathematics and Functional Programming - Thomas VanDrunen</a>
<ul>
<li>Lectures exist on the author's homepage, this book is used for a one semester university course with additional elective chapters in Graph Theory, Complexity Theory, Automata, ect.</li>
<li>List of <a href="https://math.vanderbilt.edu/schectex/commerrs/">Common Errors in Undergrad Math</a> like undistributed cancellations, sign errors, confusion about notation ect. if you missed it from the prelims section of this guide
<ul>
<li>
<em>The Art of Computer Programming Vol 1: Fundamental Algorithms</em> by Donald E. Knuth also offers an excellent but terse introduction to Discrete Math</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>Great Theoretical Ideas in Computer Science</h2>
<p>This is a crash course in multiple topics such as Probability, Linear Algebra, Modular Arithmetic, Polynomials, Cryptography and Complexity Theory. 15-251 assumes students have completed <em>An Infinite Descent into Pure Mathematics</em> though it is self contained. "This course teaches the mathematical underpinnings of computation and explores some of the central results and questions regarding the nature of computation."</p>
<h1>Computer Systems</h1>
<blockquote>
<p>"[Computer science] is not really about computers - and it's not about computers in the same sense that physics is not really about particle accelerators, and biology is not about microscopes and Petri dishes...and geometry isn't really about using surveying instruments. Now the reason that we think computer science is about computers is pretty much the same reason that the Egyptians thought geometry was about surveying instruments: when some field is just getting started and you don't really understand it very well, it's very easy to confuse the essence of what you're doing with the tools that you use." (Hal Abelson)</p>
</blockquote>
<p>This covers computer architecture from a programmer's perspective, such as how to write cache friendly code, and other optimizations for the x86-64 arch. You learn how to manually write loops in assembly and see how recursion works at a lower abstraction. You learn machine code instructions, how compilers work, return oriented programming (ROP) to bypass stack protections, the memory hierarchy, and networks. You could read K&amp;R's <em>The C Programming Language</em> for a brief intro, though this course will explain C as you go anyway and fully covers pointers at the assembly language level making it self contained.</p>
<h2>Designing Systems</h2>
<p>Knuth constructed a futuristic architecture with 256 registers for his <em>Art of Computer Programming</em> series to make hacking around with experimental systems programming easier. If you ever find yourself implementing instructions for something (webAssembly devs) you should read through <a href="http://mmix.cs.hm.edu/doc/instructions-en.html">MMIX</a> documentation, for gems like MOR/MXOR.</p>
<ul>
<li>
<a href="https://archive.org/details/springer_10.1007-3-540-46611-8">(Book) MMIXware: a RISC computer for the third millennium - Donald E Knuth</a>
<ul>
<li>An intro to MMIX <a href="http://mmix.cs.hm.edu/doc/fasc1.pdf">here</a> with some of it's interesting design (such as implementing proper integer arithmetic) and <a href="http://mmix.cs.hm.edu/getstarted.html">simulators</a>
</li>
<li>Knuth uses literate programming to go through all design decisions, describing every line of code in detail as it's how he learned to write symbolic assemblers in the 1950s by reading other people's <a href="https://github.com/kragen/knuth-interview-2006">programs</a> while on vacation</li>
<li>Combine with some <a href="https://www.cse.unsw.edu.au/~cs9242/18/lectures/">course notes</a> on open research issues for operating systems or read about the architecture and critiques of <a href="http://walfield.org/papers/200707-walfield-critique-of-the-GNU-Hurd.pdf">Hurd</a>
<ul>
<li>A <a href="https://www.morganclaypool.com/toc/cac/1/1">series</a> of papers in Computer Architecture curated by Princeton U (use <a href="https://en.wikipedia.org/wiki/Sci-Hub">sci-hub</a> to access them)</li>
</ul>
</li>
</ul>
</li>
</ul>
<h1>Compilers</h1>
<p>Search libgen.io or sci-hub for additional <a href="https://cc-conference.github.io/18/accepted-papers/">papers</a> or <a href="https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools">books</a> on parsers/compilers.</p>
<h1>Database Systems</h1>
<p>Interesting <a href="http://www.cs.cmu.edu/~./pavlo/blog/2015/09/the-next-50-years-of-databases.html">post</a> on the future of database systems by Andy Pavlo.</p>
<h1>Practical Data Science</h1>
<ul>
<li>
<a href="http://www.datasciencecourse.org/lectures/">(Full Course) 15-388 Practical Data Science</a>
<ul>
<li>Provides a practical introduction to the 'full stack' of data science analysis: data collection and processing, data visualization and presentation, statistical model building using machine learning, and big data techniques for scaling these methods.</li>
<li>Jupyter notebook is a <a href="http://www.literateprogramming.com/index.html">literate programming</a> paradigm tool, you can freely write using LaTeX or markdown and run the code cells in any order to not interrupt the flow of documentation.
<ul>
<li>This course teaches the Python libraries, but you can do the assignments in <a href="http://ucidatascienceinitiative.github.io/IntroToJulia/">Julia</a> if you want, or experiment in <a href="http://ocaml.xyz/">OCaml</a> with static typing.</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>A classic introductory computer science book on thinking about the big picture of programs with abstraction: finding general patterns from specific problems and building programs based on these patterns. An applied example of this is the package manager <a href="http://dustycloud.org/misc/talks/guix/chicagolug_2015/guix_talk.html">GNU Guix</a> and distro GuixSD, which is a GNU implementation of the NixOS <a href="https://arxiv.org/pdf/1305.4584.pdf">functional software deployment model</a>. Package builds, including entire system builds, are declared in one text file. The resulting software deployment is functional: build inputs go in such as compilers, customizations, environments ect, and a reproducible, immutable build comes out with a hashed identifier meaning you can do roll backs to previous successful builds. A recent GUixSD feature <a href="https://arxiv.org/pdf/1709.00833.pdf">gexp</a> (g-expressions) is a good example what can be achieved through syntactic extensions of the Scheme language.</p>
<p>You are more likely to benefit from this book after having some programming experience, but no matter what <a href="https://twitter.com/id_aa_carmack/status/350028210551013376?lang=en">level</a> of programmer you are you'll still benefit from SICP.</p>
<h1>Natural Language Processing</h1>
<h1>Programming Language Theory</h1>
<p>Read <a href="http://www.cs.cmu.edu/~rwh/courses/ppl/phil.html">these reasons</a> for studying programming languages.</p>
<h1>Isolating Software Failure, Proving Safety and Testing</h1>
<p>How to verify software, and strategies of programming that minimize catastrophe during failure. The <a href="https://mitpress.mit.edu/books/little-prover">Little Prover</a> is a good introduction in determining facts about computer programs. Prof Rob Simmons can be <a href="https://calculem.us/classes/">hired</a> as a tutor to teach you Coq, the curriculum follows the Software Foundations book series below.</p>
<ul>
<li>
<a href="https://15316-cmu.github.io/index.html">(Lecture Notes) 15-316 Software Foundations of Security and Privacy</a>
<ul>
<li>Covers side channel attacks, provable privacy, web security, sandboxing, assignments are extending an http network server (in C), and spotting bugs.</li>
<li>Read about tests, strategies to safe program design (in OCaml), and proving safety from the 2017 version: git clone <a href="https://github.com/jeanqasaur/cmu-15316-spring17.git">https://github.com/jeanqasaur/cmu-15316-spring17.git</a>
</li>
</ul>
</li>
<li>
<a href="http://www.cs.princeton.edu/~appel/vfa/">(Book) Verified Functional Algorithms</a>
<ul>
<li>Some recorded lectures by Andrew Appel <a href="https://deepspec.org/event/dsss17/lecture_appel.html">here</a>
</li>
<li>Part 3 of the <a href="https://deepspec.org/page/SF/">Deep Specifications</a> interactive book series by Andrew Appel, learn by doing</li>
<li>Assumes you have read <a href="https://softwarefoundations.cis.upenn.edu/current/index.html">these</a> chapters of Software Foundations Part I: Preface, Basics, Induction, Lists, Poly, Tactics, Logic, IndProp, Maps, (ProofObjects), (IndPrinciples)
<ul>
<li>A good introduction to Dependent Types by Dan Licata is <a href="https://youtu.be/LXvP1A97oAM">here</a> or The Little Typer <a href="https://mitpress.mit.edu/books/little-typer">book</a>.</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>Physical Systems Software Security</h2>
<ul>
<li>
<a href="http://symbolaris.com/course/fcps16.html">(Full Course) 15-424 Foundations of Cyber-Physical Systems</a>
<ul>
<li>Course (with recorded lectures) if you're interested in programming drones/space shuttles/robots/cars and other things that cannot fail from avoidable errors.</li>
<li>Self contained, will teach you differential eq but assumes you already have some calculus background.</li>
</ul>
</li>
</ul>
<h1>Algorithms</h1>
<h2>Introduction to Parallel and Sequential Algorithms</h2>
<p>Unlike a traditional introduction to algorithms and data structures, this puts an emphasis on thinking about how algorithms can do multiple things at once instead of one at a time and ways to design to exploit these properties. It uses a language based analysis model to estimate complexity for abstractions such as garbage collection.</p>
<h2>Advanced Algorithms (Purely Functional Data Structures)</h2>
<p>The typical text for a senior undergrad or graduate course on algorithm analysis is <em>The Design and Analysis of Algorithms</em> by Kozen (1992). This is an excellent book to study with difficult problems well presented and clearly analyzed by Kozen. The Okasaki book (1999), <em>Purely Functional Data Structures</em> gives examples in Standard ML and Haskell.</p>
<h1>Complexity Theory</h1>
<h2>Undergraduate Complexity Theory</h2>
<p>Expands on the lectures in 15-251.</p>
<h2>Graduate Complexity Theory</h2>
<p>You may want to <a href="https://news.ycombinator.com/item?id=6346697">try solving</a> some of the problems in this domain.</p>
<h2>Useful Math for Theoretical CS</h2>
<p>These scribed lecture <a href="http://www.cs.cmu.edu/~odonnell/toolkit13/">notes</a> give you a diverse background in math useful for theoretical CS, from the excellent book <a href="http://www.nature-of-computation.org/">The Nature of Computation</a> such as these <a href="http://www.cs.cmu.edu/~15751/2016-lecture1.pdf">slides</a> from <em>A Theorist's ToolKit</em> which describe how to find research, how to write math in LaTeX, how to give a talk, resources on math.overflow ect.</p>
<h1>Introduction to Quantum Computing</h1>
<p>This is a graduate introduction to quantum computation/information theory, from the perspective of theoretical computer science.</p>
<ul>
<li>
<a href="http://www.cs.cmu.edu/~odonnell/quantum18/">(Full Course) 15-859BB: Quantum Computation</a>
<ul>
<li>Recorded lectures on <a href="https://www.youtube.com/playlist?list=PLm3J0oaFux3YL5qLskC6xQ24JpMwOAeJz">YouTube</a>.</li>
<li>The prereqs are an undergrad background in complexity theory (15-251 or 15-455), linear algebra, and discrete <a href="http://www.cs.cmu.edu/~odonnell/papers/probability-and-computing-lecture-notes.pdf">probability</a>.</li>
<li>"90% of the understanding of the quantum circuit model is achieved by reviewing three purely 'classical' topics: classical Boolean circuits; reversible classical circuits; and randomized computation"
<ul>
<li>A <a href="https://www.morganclaypool.com/toc/qmc/1/1">series</a> of curated papers on Quantum Computing</li>
</ul>
</li>
</ul>
</li>
</ul>
<h1>Jobs</h1>
<p>This <a href="https://triplebyte.com/">service</a> matches your skills to people who want to pay you. Jane Street Capital is a finance tech company that hires functional programmers worldwide, you may want to <a href="https://blogs.janestreet.com/interviewing-at-jane-street/">apply</a> there. There are numerous opportunities to apprentice as a <a href="https://www.nccgroup.trust/us/about-us/careers/current-vacancies/security-consultant/">security researcher</a> or get paid to prove business logic programs like 'smart contracts'. Check your local university employment listings as well, often students do not take these jobs as they are off chasing that startup stock options payday or doing industry internships, plenty of opportunities to work with post-doc researchers writing algorithms or analyzing data. For web programming there exists free programming <a href="https://books.goalkicker.com/">notes</a> for software like NodeJS, which you can write software for in OCaml by compiling it into JS with <a href="https://bucklescript.github.io/">bucklescript</a> or compile to web assembly.</p>
<h1>Graduate Research in Type Theory</h1>
<p>Read these <a href="http://www.cs.cmu.edu/~15751/2016-lecture1.pdf">slides</a> from <em>A Theorist's ToolKit</em> on how to find research, how to write math in LaTeX, how to give a talk, where to ask on stackexchange ect.</p>
<h2>Basic Proof Theory</h2>
<ul>
<li>Start with the <a href="https://www.cs.uoregon.edu/research/summerschool/summer12/curriculum.html">lectures</a> on <em>Basic Proof Theory</em> by Frank Pfenning</li>
</ul>
<h2>Intro to Category Theory</h2>
<h2>Type Theory Foundations</h2>
<ul>
<li>Read Chapters 1-12, 23 and 24 from <a href="http://www.seas.upenn.edu/~bcpierce/tapl/index.html">TAPL</a>
</li>
<li>Watch <a href="https://www.cs.uoregon.edu/research/summerschool/summer12/curriculum.html">seminar</a> on <em>Type Theory Foundations</em> by Robert Harper</li>
</ul>
<h2>Higher Dimensional Type Theory</h2>
<p>Start with this <a href="https://youtu.be/caSOTjr1z18">talk</a> <em>A Functional Programmer's Guide to Homotopy Type Theory</em> with intro to Cubical Type Theory</p>
<h2>Further Research</h2>
<h1>Graduate Research in Machine Learning/AI</h1>
<p>Read these <a href="http://www.cs.cmu.edu/~15751/2016-lecture1.pdf">slides</a> from <em>A Theorist's ToolKit</em> on how to find research, how to write math in LaTeX, how to give a talk, where to ask on stackexchange ect.</p>
<h2>Graduate Introduction to General AI</h2>
<p>These notes <a href="http://www.cs.cmu.edu/~./15381/#schedule">here</a> cover more topics, such as Computer Vision/NLP/AGI and follows <em>AI: A Modern Approach</em> book by Norvig. The 15-780 course is specific to some topics in AI providing them 2-3 lectures each.</p>
<ul>
<li>
<a href="http://www.cs.cmu.edu/afs/cs/Web/People/15780/#schedule">(Full Course) 15-780 Graduate Introduction to AI</a>
<ul>
<li>Recorded lectures <a href="https://www.youtube.com/playlist?list=PLpIxOj-HnDsPfw9slkk0BfwuiNEYVnsd_">here</a> and <a href="https://www.youtube.com/playlist?list=PL_Ig1a5kxu55zgIn3Tb5Pc5C62Yo2sYHf">here</a>
</li>
<li>No formal pre-requisites except grad level standing so familiar with some undergrad calculus and linear algebra/probability
<ul>
<li>MIT's Artificial General Intelligence <a href="https://agi.mit.edu/">free class</a> explores possible research paths for creating AGI</li>
<li>Sussman had a great 2017 grad <a href="https://ai6034.mit.edu/wiki/index.php?title=6.S966:_A_Graduate_Section_for_6.034#Prospectus">course</a> reviewing older paper with unsolved problems or deep ideas, like his own <em>Art of the Propagator</em> paper and <a href="http://groups.csail.mit.edu/mac/users/gjs/propagators/">implementation</a>
</li>
<li>AI <a href="https://courses.csail.mit.edu/6.803/schedule.html">readings</a> to explain intelligence from a computational point of view</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>Math Background for ML</h2>
<p>A series of <a href="https://www.youtube.com/playlist?list=PL7y-1rk2cCsAqRtWoZ95z-GMcecVG5mzA">recorded</a> lectures and recommended <a href="https://canvas.cmu.edu/courses/603/assignments/syllabus">texts</a> for a crash course in the Math background for ML introductory courses</p>
<h2>Statistics Theory</h2>
<ul>
<li>
<a href="http://www.stat.cmu.edu/~larry/=stat705/">(Full Course) 36-705 Intermediate Statistics</a>
<ul>
<li>The author of <em>All of Statistics</em> Larry Wasserman covers the fundamentals of theoretical statistics in these (badly) recorded lectures</li>
<li>The probability prereqs can be found <a href="https://www.youtube.com/playlist?list=PL_Ig1a5kxu57qPZnHm-ie-D7vs9g7U-Cl">here</a> or from any undergrad text on probability</li>
<li>Additional lecture <a href="http://www.stat.cmu.edu/~larry/=sml/">notes</a> on the intersection of stats theory and ML</li>
</ul>
</li>
</ul>
<h2>Graduate Introduction to ML</h2>
<h2>Advanced Introduction to ML</h2>
<ul>
<li>
<a href="http://www.cs.cmu.edu/~bapoczos/Classes/ML10715_2015Fall/index.html">(Full Course) 10-715 Advanced Introduction to ML</a>
<ul>
<li>Most recent version <a href="https://sites.google.com/site/10715advancedmlintro2017f/lectures">here</a> uses mostly the same recorded lectures.</li>
<li>Fast paced curriculum intended to prepare PhD students in the ML grad program to write research papers</li>
<li>Some <a href="http://curtis.ml.cmu.edu/w/courses/index.php/Syllabus_for_Machine_Learning_with_Large_Datasets_10-605_in_Fall_2017">slides</a> on the practical techniques needed for working with large datasets
<ul>
<li>Linear Algebra review crash course w/recorded lectures <a href="http://www.cs.cmu.edu/~zkolter/course/linalg/index.html">here</a> relevant to optimization/ML</li>
<li>Real Analysis review <a href="https://www.youtube.com/channel/UC7gOYDYEgXG1yIH_rc2LgOw/playlists">here</a>
</li>
<li>Probability lecture <a href="http://www.stat.cmu.edu/~siva/700/main.html">notes</a>
</li>
</ul>
</li>
</ul>
</li>
<li>
<a href="http://www.cs.cmu.edu/~epxing/Class/10701/lecture.html">(Full Course) 10-701 Introduction to ML</a>
<ul>
<li>Intended for PhD students outside the ML program, more theory and rigor than 10-601</li>
<li>Assumes you have this <em>Math Background for ML</em> math background</li>
<li>Some recorded recitations are <a href="https://www.youtube.com/playlist?list=PL_Ig1a5kxu54nrfuy4V0V1eHJk3C4dV3X">here</a>
<ul>
<li>Some <a href="http://curtis.ml.cmu.edu/w/courses/index.php/Syllabus_for_Machine_Learning_with_Large_Datasets_10-605_in_Fall_2017">slides</a> on the practical techniques needed for working with large datasets</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>Convex Optimization</h2>
<h2>Deep Learning</h2>
<p>If you're interested in parallel GPU programming for training see these <a href="https://www.youtube.com/playlist?list=PLpIxOj-HnDsO4Atvrp86c-4La9Mq3kMQZ">lectures</a> and <a href="http://15418.courses.cs.cmu.edu/spring2016/lectures">notes</a>.</p>
<ul>
<li>
<a href="https://www.cs.cmu.edu/~bhiksha/courses/deeplearning/Fall.2015/">(Full Course) 11-785 Introduction to Deep Learning</a>
<ul>
<li>Recorded lectures <a href="https://www.youtube.com/playlist?list=PLwrNrH82f0KgdZhGB7sm9NrOanfWBS-EN">here</a> to learn to build and tune deep learning models
<ul>
<li>If the playlist is deleted, which is frequently, (use youtube-dl to archive) search YouTube for "CMU 11-785"</li>
</ul>
</li>
<li>There are also applied courses and practical challenges/competitions on kaggle.com
<ul>
<li>Some <a href="http://www.archive.ece.cmu.edu/~ece739/schedule.html">slides</a> on techniques for making deep learning robust to adversarial manipulation</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>Algorithms for Big Data</h2>
<h2>Further Research</h2>
<h1>Graduate Research Elective: Cryptography</h1>
<p>Read these <a href="http://www.cs.cmu.edu/~15751/2016-lecture1.pdf">slides</a> from <em>A Theorist's ToolKit</em> on how to find research, how to write math in LaTeX, how to give a talk, where to ask on stackexchange ect.</p>
<h2>Graduate Cryptography Intro</h2>
<p>This course covers post-quantum crypto, elliptic curve crypto, and is taught by premiere researcher Tanja Lange (TU/e)</p>
<h2>Applied Cryptography</h2>
<ul>
<li>Draft <a href="https://crypto.stanford.edu/~dabo/cryptobook/">book</a> <em>A Graduate Course in Applied Cryptography</em> - Dan Boneh and Victor Shoup</li>
<li>Lecture <a href="https://www.ece.cmu.edu/~ece733/schedule.html">notes</a> from 18-733 Applied Cryptography</li>
<li>Read all Daniel J. Bernstein's (and Peter Gutmann's) posts on the IETF Crypto Forum Research Group [Cfrg] archive, it's a <a href="https://mailarchive.ietf.org/arch/msg/cfrg/E-q_6ABdKfPU5XG2N6IVAs_sNhA/?qid=435ba2be2c3db3444e338cc0259ee124">master class</a> in modern <a href="https://mailarchive.ietf.org/arch/msg/cfrg/oOY_QNA6QMsTJHLLbUjch6-GLsc/?qid=435ba2be2c3db3444e338cc0259ee124">cryptanalysis</a> and he rips apart bad standards/protocol designs.</li>
<li>Read <em>The Art of Computer Programming - Seminumerical Algorithms</em> by Knuth (Vol 2) chapter on Random Numbers. These tests are still used in MIT grad courses. Try them on every library you can find that supposedly generates pseudorandom numbers</li>
<li>
<a href="https://www.wireguard.com/formal-verification/">Read</a> about the proof of the Wireguard protocol, a VPN that uses <code>AEAD_CHACHA20_POLY1305</code>
</li>
</ul>
<h2>Future Research</h2>
<ul>
<li>Follow whatever the PhD students of djb and Tanja Lange are <a href="https://hyperelliptic.org/tanja/students.html">working on</a>
</li>
<li>
<a href="https://2017.pqcrypto.org/school/schedule.html">Watch</a> the lectures from the 2017 Post-Quantum Crypto Summer School</li>
<li>Read the journal of <a href="https://link.springer.com/journal/13389">Crypto Engineering</a> (use SciHub proxy)</li>
<li>Read a <a href="http://www.math.cmu.edu/~af1p/BOOK.pdf">book</a> on Random Graphs there is a <a href="http://math.nsc.ru/conference/g2/g2c2/TokarevaN.pdf">connection</a> between Graph Theory and Cryptography</li>
<li>Try the <a href="https://cryptopals.com/">Cryptopals</a> challenges</li>
<li>Read some cryptocurrency papers, such as Stellar's Consensus algorithm (soon to be used by mobilecoin.com), <a href="https://medium.com/@homakov/introducing-failsafe-network-ea47ab476fe6">Fail-Safe Network</a> or the <a href="https://github.com/zcash/zips/blob/master/protocol/protocol.pdf">protocol specification</a> for Zcash.</li>
</ul>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Site not found &middot; GitHub Pages<br>
<br>
January 31, 2019 at 11:11AM
